Refactor to test gdppc, decouple data model

brews opened the pull request:
Refactors `impactcommon/exogenous_economy/gdppc.py:GDPpcProvider` into `HierarchicalGDPpcProvider`, extracting dependence on file IO and directory structure into `read_bestgdppcprovider()`. In short, these changes make it easy to test the gdppc.py components. This also  helps decouple the domain model from the data model.

- `GDPpcProvider` is now a function to support legacy `GDPpcProvider` class behavior.
- The provider class was renamed `HierarchicalGDPpcProvider`.
- Removes globals variables `growth_filepath`, `baseline_filepath`, `nightlights_filepath`, `baseline_year`. Replaced with default arguments. This avoids shared, hidden states.
- All new tests run fast and run in CI/CD.
- New documentation.

Close #19

brews committed:
Decouple data IO, input cleaning, globals from GDPpc provider

This will make things cleaner and testing easier.

brews committed:
Minor cleanup to bad sharedpath pass in GDPpcProvider

brews committed:
Refactor to add gdppc.py testing

brews committed:
Minor simplification to test inputs

brews committed:
Move test data creation to fixtures

brews committed:
Minor docstr fix

brews committed:
Refactor signature, growth yearindex to GdpProvider for clarity

brews committed:
Consolidate test input setup further

brews committed:
Rename GdpProvider to BestGDPpcProvider and with sup functions

brews committed:
Minor style cleanup

brews committed:
Remove unused import

brews committed:
Simplify, swapping handmade cache with lru_cache

brews committed:
Minor style cleanup

brews committed:
Improve docstrs

brews committed:
Parametrize test

brews committed:
Remove TODO

brews commented:
This PR also passed all downstream package system tests.

jrising commented on impactcommon/exogenous_economy/gdppc.py:
```diff
@@ -1,55 +1,213 @@
-baseline_year = 2010
 
-class GDPpcProvider(provider.BySpaceProvider):
+def read_bestgdppcprovider(iam, ssp, growth_path_or_buffer, baseline_path_or_buffer, nightlights_path_or_buffer,
```
Conceptually, I think this would make more sense as a static or class method of `BestGDPpcProvider`.

brews replied:
@jrising This is pretty clean. Why bring more unneeded OOP into this?

jrising replied:
One solution would be to make this a clearinghouse function. Rather than saying that it's only ever going to return `BestGDPpcProvider` objects (or whatever we call them), it could return provide a one-stop-shop for whatever the default `GDPpcProvider` is (currently set to `BestGDPpcProvider`). Then, when we make a new one, but still want to provide access to the old one for reproducing past results, we can let it take a configuration option. But that all seems more convoluted than just letting people instantiate whichever class they want, using a helper function defined on that class.

brews replied:
@jrising That's something to consider but this is all overengineering until we have more than one `*Provider`.

brews replied:
We discussed this at last checkin. Marking resolved.

jrising commented on impactcommon/exogenous_economy/gdppc.py:
```diff
@@ -1,55 +1,213 @@
-        self.df_baseline_this = df.loc[(df.model == iam) & (df.scenario == ssp) & (df.year == baseline_year)]
-        self.df_baseline_anyiam = df.loc[(df.scenario == ssp) & (df.year == baseline_year)].groupby('iso').median()
-        self.baseline_global = df.loc[(df.scenario == ssp) & (df.year == baseline_year)].median()
+    out = BestGDPpcProvider(
```
Why define the `out` variable if you're not going to use it?

brews replied:
No problem. Fixed in `844d376a`.

jrising commented on impactcommon/exogenous_economy/gdppc.py:
```diff
@@ -1,55 +1,213 @@
+    return out
+
+
+def _get_best_iso_available(iso, df_this, df_anyiam, df_global):
```
It seems cluttering to put this in the module-scope. This function only exists as a step in the `BestGDPpcProvider` algorithm, so it should be bundled as a static method with that object.

brews replied:
@jrising You're cluttering your class interface with logic that could have been a simple function. Don't use static methods when a dumb function will do.

jrising replied:
Imagine that someone in the future wants to read this file and try to understand it. If you just drop this here, they have to figure out where it's used. If in the future there are more implementations of `GDPpcProvider` defined here, they won't know if they can change this without trying to understand the logic of all of the other classes in the file.

One of the biggest reasons for OOP is encapsulation. And all of python is OO, so you're always choosing a level of encapsulation. Here, you've chosen to associate this function with a wider scope than necessary.

brews replied:
Python being object-orient doesn't mean that it's good design to make everything a bunch of god classes. It just means that everything is an object (including functions).

Choosing to do this with a a static method or a function doesn't change the level of encapsulation because either one is abstracted away from how users interact with the class.

On one has you have someone like Luciano Ramalho who argues that static method is completely useless, on the other hand you have someone like Leonardo Rochael, who argues that it's useful because you can have child classes modify behavior (I'd agree that this is a useful case). Nobody argues that a python best practice is to make a static method simply because it has some relation to the object. This is not a good practice, ever.

brews replied:
Addressed this in `d774eb3e`. This removes the `_get_best_iso_available()` function and its tests and sticks `_get_best_iso_available()`  back into a method.

jrising commented on impactcommon/exogenous_economy/gdppc.py:
```diff
@@ -1,55 +1,213 @@
+    return df_global
+
+
+def _split_baseline(df, iam, ssp, baseline_year):
```
I don't see the utility in defining these as separate functions, unless you're going to have them provide a reusable service. My suggestion is to do the following:
1. Merge these into a single `_split_hierarchy` function, with logic identical to the current `_split_growth`.
2. Make this a method of `HierarchicalGDPpcProvider`.
3. Use it twice, but when used in the place of `_split_baseline`, first subset by year.

brews replied:
Addressed this with `e9182ef2`, and `a476380d`.

I removed the split tests and reverted it to its original state.

Your suggested merge doesn't quite produce the same results in the output Dataframe. The "iso" and "year" change position between the column and the index. This means we'd have to make changes elsewhere. Revisit this later once tests from this PR are in place?

jrising replied:
That's fine.

jrising commented on impactcommon/exogenous_economy/gdppc.py:
```diff
@@ -1,55 +1,213 @@
+
+class BestGDPpcProvider(provider.BySpaceProvider):
+    """
+    Provider of GDP per capita (GDPpc) timeseries, selecting "best" available source
```
I know I used the word "best" once in the original code, but I don't think we should make a habit of it. Maybe replace "best" here and in the next line with "most good"?

brews replied:
@jrising, yeah, looking back on this, I can see how "Best" could become a problem. Trying another tack: what about going simpler with just `GdpProvider()`?

We don't have any other GDP-serving providers in use?

jrising replied:
That would imply that it gives out GDPs. This doesn't even know the population.

brews replied:
Ug, yeah. I should have seen that. Stupid question on my part. Implemented your suggestion of `Hierarchical*` in `aea2ee97`.

jrising commented on impactcommon/exogenous_economy/gdppc.py:
```diff
@@ -62,50 +220,39 @@ def get_timeseries(self, hierid):
         
         return iso_gdppcs * ratio.values[0]
 
+    @lru_cache(maxsize=None)
```
LRU cache has to do some sorting on every call, which is wasted time with no maxsize. (And we definitely don't want a max size, since we know there will only ever be ~200 entries). All we really want is a one-step lookup memoization. There must be something simpler for that (or we could make that decorator).

brews replied:
@jrising, I don't think that's how this is implemented. When maxsize is None, this is simply unbounded cache: https://github.com/python/cpython/blob/4b47a5b6ba66b02df9392feb97b8ead916f8c1fa/Modules/_functoolsmodule.c#L828 

It's a C-wrapper. It adjusts references (this is how python does garbage collection) and counts hits and misses and that's about it.

`@lru_cache(maxsize=None)` becomes `@cache`, in later releases.

jrising replied:
Okay, that's fine then.

jrising commented on tests/test_gdppc.py:
```diff
@@ -1,59 +1,325 @@
-    )
+import numpy as np
+import pandas as pd
 from impactcommon.exogenous_economy import gdppc
```
Can you add back in the "real data" tests, for when we have the shared dir available?

brews replied:
With `69ecf951` the old tests are now in `tests/test_gdppc_externaldata.py`. They remain in their own module so that they can be marked and then skipped if external data is not available.

jrising replied:
That works. Thanks.

jrising reviewed (changes requested):
This is definitely an improvement. Thanks for putting it together. I have issues with some of the design (as commented), but I appreciate what you're doing.

I think `Best` is a confusing name here. What could we possibly call the next class we make to perform this, after we improve on `Best`? I know that's not how you mean it, but better to use something like `HierarchicalGDPpcProvider`.

brews committed:
Minor style cleanup in tests

brews committed:
Refactor data splits back into class methods, remove split tests

This addresses reviewer feedback.

brews committed:
Refactor _get_best_iso_available back into method, drop function tests

This is to address reviewer comments.

brews committed:
Minor style cleanup

brews committed:
Style cleanup to address reviewer comments

brews committed:
Rename bestgdppcprovider to hierarchicalgdppcprovider

Addresses reviewer comment

brews committed:
Reinsert old unit tests at reviewer request

brews committed:
Remove cruft

brews commented:
@jrising I think I addressed most of your concerns here. I pulled back the scope of the PR to focus on separating IO and the provider class (without breaking legacy) and getting this basic separation tested. Let's get this PR done and we can fuss over the other bits later.

I think this is ready for re-review, when you have a free moment. Thanks! 👍

jrising commented on impactcommon/exogenous_economy/gdppc.py:
```diff
@@ -1,55 +1,213 @@
 
-        super(GDPpcProvider, self).__init__(iam, ssp, baseline_year)
+    df = metacsv.read_csv(baseline_path_or_buffer)
+    df_growth = metacsv.read_csv(files.sharedpath(growth_path_or_buffer))
```
Do you mean for these calls to `sharedpath` to still be here?

brews replied:
Ah. That's a bug.

Thank you! Fixed with `842f05ba`

jrising reviewed (approved):
Great. I'm glad to have the data separated out.

brews committed:
Fix bug using sharedpath when shouldn't
